llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -104,59 +111,101 @@ StringRef getStringLiteralArg(const CallInst *CI, unsigned ArgNo,
-  switch (T->getIntegerBitWidth()) {
-  case 1:
-    return "b";
-  case 8:
-    return "a";
-  case 16:
-    return "s";
-  case 32:
-    return "i";
-  case 64:
-    return "x";
-  default:
-    llvm_unreachable("unsupported spec const integer type");
+  if (T->isIntegerTy()) {
+    switch (T->getIntegerBitWidth()) {
+    case 1:
+      return "b";
+    case 8:
+      return "a";
+    case 16:
+      return "s";
+    case 32:
+      return "i";
+    case 64:
+      return "x";
+    default:
+      llvm_unreachable("unsupported spec const integer type");
+    }
   }
+  // Mangling, which is generated below is not conformant with C++ ABI rules
+  // (https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangle.unqualified-name)
+  // But it should be more or less okay, because these declarations only
+  // exists in the module between invocations of sycl-post-link and llvm-spirv,
+  // llvm-spirv doesn't care about the mangling and the only intent here is to
+  // make sure that we won't encounter redefinition error when we proceed two
+  // spec constants with different types.
+  if (T->isStructTy()) {
+    return T->getStructName().str();
+  }
+  if (T->isArrayTy()) {
+    return "A" + manglePrimitiveType(T->getArrayElementType());
+  }
+  if (auto *VecTy = dyn_cast<FixedVectorType>(T)) {
+    return "Dv" + std::to_string(VecTy->getNumElements()) + "_" +
+           manglePrimitiveType(VecTy->getElementType());
+  }
+  llvm_unreachable("unsupported spec const type");
   return "";
 }
 
 // This is a very basic mangler which can mangle non-templated and non-member
 // functions with primitive types in the signature.
-std::string mangleFuncItanium(StringRef BaseName, FunctionType *FT) {
+std::string mangleFuncItanium(StringRef BaseName, const FunctionType *FT) {
   std::string Res =
       (Twine("_Z") + Twine(BaseName.size()) + Twine(BaseName)).str();
   for (unsigned I = 0; I < FT->getNumParams(); ++I)
     Res += manglePrimitiveType(FT->getParamType(I));
   return Res;
 }
 
-void setSpecConstMetadata(Instruction *I, StringRef SymID, int IntID) {
+void setSpecConstSymIDMetadata(Instruction *I, StringRef SymID,
+                               ArrayRef<unsigned> IntIDs) {
   LLVMContext &Ctx = I->getContext();
-  MDString *SymV = MDString::get(Ctx, SymID);
-  ConstantAsMetadata *IntV =
-      ConstantAsMetadata::get(ConstantInt::get(Ctx, APInt(32, IntID)));
-  MDNode *Entry = MDNode::get(Ctx, {SymV, IntV});
+  SmallVector<Metadata *, 4> MDOperands;
+  MDOperands.push_back(MDString::get(Ctx, SymID));
+  for (unsigned ID : IntIDs) {
+    MDOperands.push_back(
+        ConstantAsMetadata::get(ConstantInt::get(Ctx, APInt(32, ID))));
+  }
+  MDNode *Entry = MDNode::get(Ctx, MDOperands);
   I->setMetadata(SPEC_CONST_SYM_ID_MD_STRING, Entry);
 }
 
-std::pair<StringRef, unsigned> getSpecConstMetadata(Instruction *I) {
+std::pair<StringRef, unsigned>
+getScalarSpecConstMetadata(const Instruction *I) {
```

Comment:
> Is it some clang-format change?

Yes, this code is properly formatted

> Does it make sense to define an alias for `std::pair<StringRef, unsigned>` type?

I'm not sure, because I can't think of a shorter name which would represent that pair and also this pair type is only mentioned once.